### PR TITLE
Pointer captures + touch scroll 

### PR DESCRIPTION
**Pull request by kekekeks**
https://www.youtube.com/watch?v=bIkyLdph2LA

### CODE REVIEW

File: src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs [hunk 1]
```diff
@@ -227,6 +230,72 @@ protected override Size ArrangeOverride(Size finalSize)
             return finalSize;
         }
 
+        // Arbitrary chosen value, probably need to ask ILogicalScrollable
+        private const int LogicalScrollItemSize = 50;
```
Comment (kekekeks) [inline src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs]: Note: this value was taken from mouse wheel scrolling: https://github.com/AvaloniaUI/Avalonia/blob/551ce9b058afc572f112252dd8ea8bb44cdd8301/src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs#L312

File: src/Avalonia.Controls/TabControl.cs [hunk 2]
```diff
@@ -166,10 +168,24 @@ protected override void OnPointerPressed(PointerPressedEventArgs e)
...
             base.OnPointerPressed(e);
 
-            if (e.MouseButton == MouseButton.Left)
+            if (e.MouseButton == MouseButton.Left && e.Pointer.Type == PointerType.Mouse)
             {
                 e.Handled = UpdateSelectionFromEventSource(e.Source);
             }
         }
+
+        protected override void OnPointerReleased(PointerReleasedEventArgs e)
```
Comment (kekekeks) [inline src/Avalonia.Controls/TabControl.cs]: Note: mouse release behavior is needed to allow scroll gesture recognizer to take over before TabControl switches the tab

**Review by grokys — CHANGES_REQUESTED** [review]
Just tried this on my Dell XPS13 (Win10) and the touch position is wrong: it's offset from where I'm touching:

https://www.youtube.com/watch?v=bDmaaFTc5qE

File: src/Avalonia.Input/MouseDevice.cs [hunk 3]
```diff
@@ -34,27 +30,9 @@ public class MouseDevice : IMouseDevice, IPointer
...
-
-                if (value != null)
-                {
-                    _capturedSubscription = Observable.FromEventPattern<VisualTreeAttachmentEventArgs>(
-                        x => value.DetachedFromVisualTree += x,
-                        x => value.DetachedFromVisualTree -= x)
-                        .Take(1)
-                        .Subscribe(_ => Captured = null);
-                }
+        [Obsolete("Use IPointer instead")]
```
Comment (grokys) [inline src/Avalonia.Input/MouseDevice.cs]: Might be clearer as `Use IPointer.Capture instead`. Also, how does one get an `IPointer` for a `MouseDevice` now? `MouseDevice` no longer implements `IPointer` and  `_pointer` is private.

File: src/Avalonia.Input/Pointer.cs [hunk 4]
```diff
@@ -9,23 +11,40 @@ public class Pointer : IPointer, IDisposable
...
             Type = type;
             IsPrimary = isPrimary;
-            ImplicitlyCaptured = implicitlyCaptured;
-            if (ImplicitlyCaptured != null)
-                ImplicitlyCaptured.DetachedFromVisualTree += OnImplicitCaptureDetached;
         }
 
         public int Id { get; }
 
+        IInputElement FindCommonParent(IInputElement control1, IInputElement control2)
```
Comment (grokys) [inline src/Avalonia.Input/Pointer.cs]: There is already `VisualExtensions.FindCommonVisualAncestor`: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/VisualTree/VisualExtensions.cs#L46 - does this not work properly?

File: src/Avalonia.Input/GestureRecognizers/ScrollGestureRecognizer.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
+using System;
+using System.Diagnostics;
+using Avalonia.Interactivity;
+using Avalonia.Threading;
+
+namespace Avalonia.Input.GestureRecognizers
+{
+    public class ScrollGestureRecognizer 
+        : StyledElement, // It's not an "element" in any way, shape or form, but TemplateBinding refuse to work otherwise
```
[the remark this answers is not in the document]
Comment (kekekeks) [reply src/Avalonia.Input/GestureRecognizers/ScrollGestureRecognizer.cs]: TemplateBinding's _parent is `StyledElement`. It refuses to use anything else.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup/Data/TemplateBinding.cs#L20

File: src/Avalonia.Input/MouseDevice.cs (hunk 3 above)
Comment (kekekeks) [reply src/Avalonia.Input/MouseDevice.cs]: > how does one get an IPointer for a MouseDevice

I don't think there is a valid scenario when one needs to manipulate an *arbitrary* pointer capture outside of event handling procedure. When user interacts with the app using touch input capturing mouse pointer won't have any usable effect.

So if one want to capture a pointer, said pointer should be passed/saved from PointerEventArgs.

File: src/Avalonia.Input/Pointer.cs (hunk 4 above)
Comment (kekekeks) [reply src/Avalonia.Input/Pointer.cs]: It lacks `OfType<IInputElement>()`, since we aren't interested in just visuals.

**Comment by kekekeks** [comment]
@grokys  Fixed DPI handling

File: src/Avalonia.Input/GestureRecognizers/ScrollGestureRecognizer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Input/GestureRecognizers/ScrollGestureRecognizer.cs]: Yeah. need to check how WPF/UWP handle this first, and try to relax that requirement if possible.

**Review by grokys — APPROVED** [review]
Looks good, and works great! We could do with some documentation but that's never stopped us before so 👍

**Pull request merged**